✨ [Attachment Forms] Create status indicators that display the page attachment form's submission status

coreymasanto opened the pull request:
The `amp-form` extension allows for the usage of 3 response attributes for rendering the status of a form submission attempt: `submitting`, `submit-success`, and `submit-error`. Using these attributes, we are adding default status indicators to be used for every page attachment form. Publishers have the option to override a default indicator by including an element with the desired response attribute within their form.

**submitting state**
<img src="https://user-images.githubusercontent.com/2376601/133043810-4999ff3d-3da0-48d4-a2e6-85ce31c25580.gif" width="235px">
<br>

**submit-success state**
<img src="https://user-images.githubusercontent.com/2376601/133041234-143512a0-caf9-445d-a084-834968089816.png" width="235px">
<br>

**submit-error state**
<img src="https://user-images.githubusercontent.com/2376601/133041152-5fc1c1be-76f4-48fe-9fa0-112400915d54.png" width="235px">
<br>

**Full example**
<img src="https://user-images.githubusercontent.com/2376601/133118469-6a9020e1-d05d-454b-9468-37722881f3bb.gif" width="235px">

#35569

/cc @gmajoulet

coreymasanto commented:
Hey @gmajoulet, mind adding yourself as reviewer?

gmajoulet commented on examples/amp-story/attachment.html:
```diff
@@ -174,8 +181,14 @@
+            <form
+              method="POST"
+              action-xhr="https://www.example.com"
+              target="_top"
```
Do we have to enforce this is always `_top` and never `_blank`? What happens if you specify `_blank`?

coreymasanto replied:
`_blank` supposedly opens the submission response in a new window or tab, which may not be a flow we want to allow.

I haven't been able to test amp-form with a valid endpoint (because my local server uses HTTP and `amp-form` requires and HTTPS action), so I can't yet confirm the behavior of `_blank` within a story. I'll observe the difference in behavior when I create my Firebase demo, and then we can enforce `_top` if necessary

gmajoulet commented on extensions/amp-story/1.0/amp-story-page-attachment.css:
```diff
@@ -125,3 +125,87 @@ amp-story[desktop] .i-amphtml-story-page-attachment-remote  {
+
+.i-amphtml-story-page-attachment-form-submission-status {
+  align-items: center !important;
+  background: #FFFFFF !important;
```
if you use `--i-amphtml-draggable-drawer-background-color` you should be able to get rid of the theme override

coreymasanto replied:
Updated, although the override still seems necessary for the dark theme's border

gmajoulet commented on extensions/amp-story/1.0/amp-story-page-attachment.css:
```diff
@@ -125,3 +125,87 @@ amp-story[desktop] .i-amphtml-story-page-attachment-remote  {
+.i-amphtml-story-page-attachment-form-submission-status div {
+  /* The entire [submitting] div will fade in along with its border, after which only the contents
+  of the [submit-success] and [submit-error] divs will fade in. */
+  animation: submission-status-fade-in 0.3s !important;
```
Plz use decelerated easing to show the div, and accelerated easing to hide it (if needed)
https://material.io/design/motion/speed.html#easing

.2s to show, .15s to hide should also work better given how small the element is

coreymasanto replied:
Updated, thanks for the link!

gmajoulet commented on extensions/amp-story/1.0/amp-story-page-attachment.css:
```diff
@@ -125,3 +125,87 @@ amp-story[desktop] .i-amphtml-story-page-attachment-remote  {
+}
+
+[theme="dark"] .i-amphtml-story-page-attachment-form-submission-status .i-amphtml-story-spinner-layer {
+  color: white !important;
```
Is there any CSS variable you can use to remove this? Like the font-color maybe?

coreymasanto replied:
Yup, font-color works! Didn't realize the variables could be reused between files :)

gmajoulet commented on extensions/amp-story/1.0/amp-story-page-attachment.js:
```diff
@@ -52,6 +54,15 @@ const AttachmentType = {
+/**
+ * @enum {string}
+ */
+const FormResponseAttribute = {
```
Should we import `amp-form.FormState` instead of duplicating it?

coreymasanto replied:
I'm leaning towards not using `FormState` because it represents the 6 possible [states](https://amp.dev/documentation/components/amp-form/#classes-and-css-hooks) of the form, whereas the `FormResponseAttribute` represents the 3 attributes used for rendering responses in the form

gmajoulet commented on extensions/amp-story/1.0/amp-story-page-attachment.js:
```diff
@@ -148,12 +159,17 @@ export class AmpStoryPageAttachment extends DraggableDrawer {
       this.headerEl.append(this.createDomainLabelElement_());
       this.headerEl.classList.add('i-amphtml-story-page-attachment-with-form');
+      this.storeService.dispatch(Action.ADD_TO_ACTIONS_ALLOWLIST, [
+        {tagOrTarget: 'FORM', method: 'submit'},
```
What other actions do we need to support? Maybe `clear`?
https://amp.dev/documentation/components/amp-form/#actions

coreymasanto replied:
Updated to support `clear`. I believe these two should be the only ones we need to support

gmajoulet commented on extensions/amp-story/1.0/amp-story-page-attachment.js:
```diff
@@ -483,4 +489,94 @@ export class AmpStoryPageAttachment extends DraggableDrawer {
     const publisherOrigin = getSourceOrigin(this.getAmpDoc().getUrl());
     return publisherOrigin.replace(/https?:\/\//, '');
   }
+
```
Wdyt about moving these to a `amp-story-form.js` file?

In terms of structure it could look like this file: https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/semantic-render.js

It'd make any amp-form related code easier to re-use. Also, it's not specific to amp-story-page-attachment at all.

coreymasanto replied:
Updated the PR with a new `amp-story-form.js` file

coreymasanto commented:
I've updated the `submitting` element so that it always scrolls into view when it is shown. I don't do the same for the `submit-success` and `submit-error` states, in order to avoid the page attachment scrolling twice in quick succession.

<img src="https://user-images.githubusercontent.com/2376601/133575070-da23ea35-323c-4587-b71f-4febd204de12.gif" width="200px">

coreymasanto commented:
> I've updated the `submitting` element so that it always scrolls into view when it is shown. I don't do the same for the `submit-success` and `submit-error` states, in order to avoid the page attachment scrolling twice in quick succession.

I encountered an Android issue where the `submitting` element was not being scrolled into view whenever one of the form's inputs had focus (note: calling `blur()` did not help). However, the `submit-success` and `submit-error` elements seem to have no issue being scrolled into view. This is potentially because of behavioral differences (e.g., the `submitting` element is generally only fleetingly shown, it contains an animated icon, it is the first one shown after pressing submit while an input has focus, etc).

So, I've updated the logic to scroll each of these elements into view when they are displayed.

gmajoulet commented on extensions/amp-story/1.0/amp-story-form.css:
```diff
@@ -0,0 +1,80 @@
+
+
```
```suggestion
```

gmajoulet commented on extensions/amp-story/1.0/amp-story-form.js:
```diff
@@ -0,0 +1,146 @@
+ *     added.
+ * @private
+ */
+export function setupResponseAttributeElements(drawerEl, formEl) {
```
Passing the `drawerEl` here makes this function impossible to re-use in other contexts.
I think passing `win` should allow you to do all you need. Given you haven't been able to get `scrollIntoView` to work for all cases, it'd be OK to drop it.

coreymasanto replied:
I updated this function so that `win` is passed, and I updated the scrolling logic so that it is instead set up from within the `amp-story-page-attachment.js`. I think we can keep the scrolling logic to help prevent our forms from appearing unresponsive.

gmajoulet commented on extensions/amp-story/1.0/amp-story-form.js:
```diff
@@ -0,0 +1,146 @@
+ * @param {!DraggableDrawer} drawerEl The draggable drawer AMP element to which
+ *     the form belongs.
+ */
+export function allowlistFormActions(drawerEl) {
```
passing `win` to make it more re-usable would be cool

gmajoulet commented:
@ampproject/wg-performance @coreymasanto can any of you look into why this is adding 0.82kB to the bundle-size? That looks very high for the number of lines this PR has

jridgewell commented:
> That looks very high for the number of lines this PR has

It seems correct to me. You're inlining a CSS file (with long svgs), and the regular JS code.
